File: federation-memory/fed-memory.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const { randomUUID } = require('crypto');

const PORT = 3002;
const DATA_DIR = path.join(__dirname, 'data');
const MEMORY_FILE = path.join(DATA_DIR, 'memories.json');

if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });

function loadMemories() {
  try {
    if (fs.existsSync(MEMORY_FILE)) {
      return JSON.parse(fs.readFileSync(MEMORY_FILE, 'utf8'));
    }
  } catch (e) {}
  return [];
}

function saveMemories(memories) {
  fs.writeFileSync(MEMORY_FILE, JSON.stringify(memories, null, 2));
}

function parseBody(req) {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => { try { resolve(JSON.parse(body)); } catch { resolve({}); } });
    req.on('error', reject);
  });
}

function json(res, data, status = 200) {
  res.writeHead(status, { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' });
  res.end(JSON.stringify(data));
}

function sendDashboard(req, res) {
  const dashPath = path.join(__dirname, 'memory-dashboard.html');
  if (fs.existsSync(dashPath)) {
    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(fs.readFileSync(dashPath, 'utf8'));
  } else {
    res.writeHead(404);
    res.end('Dashboard not found');
  }
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url, `http://localhost:${PORT}`);

  if (req.method === 'OPTIONS') {
    res.writeHead(200, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET,POST,DELETE,OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type'
    });
    return res.end();
  }

  if (url.pathname === '/' && req.method === 'GET') return sendDashboard(req, res);

  if (url.pathname === '/api/memory/save' && req.method === 'POST') {
    const body = await parseBody(req);
    const memories = loadMemories();
    const memory = {
      id: randomUUID(),
      content: body.content || '',
      tags: body.tags || [],
      timestamp: body.timestamp || new Date().toISOString(),
      created: new Date().toISOString()
    };
    memories.push(memory);
    saveMemories(memories);
    return json(res, { success: true, memory });
  }

  if (url.pathname === '/api/memory/list' && req.method === 'GET') {
    return json(res, loadMemories());
  }

  if (url.pathname === '/api/memory/retrieve' && req.method === 'GET') {
    let memories = loadMemories();
    const tag = url.searchParams.get('tag');
    const text = url.searchParams.get('text');
    const from = url.searchParams.get('from');
    const to = url.searchParams.get('to');

    if (tag) memories = memories.filter(m => m.tags.includes(tag));
    if (text) memories = memories.filter(m => m.content.toLowerCase().includes(text.toLowerCase()));
    if (from) memories = memories.filter(m => m.timestamp >= from);
    if (to) memories = memories.filter(m => m.timestamp <= to);

    return json(res, memories);
  }

  if (url.pathname.startsWith('/api/memory/') && req.method === 'DELETE') {
    const id = url.pathname.split('/api/memory/')[1];
    let memories = loadMemories();
    const before = memories.length;
    memories = memories.filter(m => m.id !== id);
    saveMemories(memories);
    return json(res, { success: true, deleted: before - memories.length });
  }

  if (url.pathname === '/api/memory/export' && req.method === 'POST') {
    return json(res, { success: true, memories: loadMemories() });
  }

  if (url.pathname === '/api/memory/import' && req.method === 'POST') {
    const body = await parseBody(req);
    const existing = loadMemories();
    const imported = body.memories || [];
    const merged = [...existing, ...imported];
    saveMemories(merged);
    return json(res, { success: true, imported: imported.length, total: merged.length });
  }

  json(res, { error: 'Not found' }, 404);
});

server.listen(PORT, () => {
  console.log(`Federation Memory Server running on http://localhost:${PORT}`);
  console.log(`Dashboard: http://localhost:${PORT}/`);
});

File: federation-memory/fed-memory.py
import http.server
import json
import os
import uuid
from urllib.parse import urlparse, parse_qs
from datetime import datetime

PORT = 3002
DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data')
MEMORY_FILE = os.path.join(DATA_DIR, 'memories.json')

os.makedirs(DATA_DIR, exist_ok=True)

def load_memories():
    try:
        if os.path.exists(MEMORY_FILE):
            with open(MEMORY_FILE, 'r') as f:
                return json.load(f)
    except:
        pass
    return []

def save_memories(memories):
    with open(MEMORY_FILE, 'w') as f:
        json.dump(memories, f, indent=2)

class MemoryHandler(http.server.BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET,POST,DELETE,OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def send_json(self, data, status=200):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def read_body(self):
        length = int(self.headers.get('Content-Length', 0))
        if length:
            return json.loads(self.rfile.read(length))
        return {}

    def do_GET(self):
        parsed = urlparse(self.path)

        if parsed.path == '/':
            self.send_response(200)
            self.send_header('Content-Type', 'text/html')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            dash = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'memory-dashboard.html')
            with open(dash, 'r') as f:
                self.wfile.write(f.read().encode())
            return

        if parsed.path == '/api/memory/list':
            return self.send_json(load_memories())

        if parsed.path == '/api/memory/retrieve':
            params = parse_qs(parsed.query)
            memories = load_memories()
            tag = params.get('tag', [None])[0]
            text = params.get('text', [None])[0]
            fr = params.get('from', [None])[0]
            to = params.get('to', [None])[0]
            if tag:
                memories = [m for m in memories if tag in m.get('tags', [])]
            if text:
                memories = [m for m in memories if text.lower() in m.get('content', '').lower()]
            if fr:
                memories = [m for m in memories if m.get('timestamp', '') >= fr]
            if to:
                memories = [m for m in memories if m.get('timestamp', '') <= to]
            return self.send_json(memories)

        self.send_json({'error': 'Not found'}, 404)

    def do_POST(self):
        parsed = urlparse(self.path)

        if parsed.path == '/api/memory/save':
            body = self.read_body()
            memories = load_memories()
            memory = {
                'id': str(uuid.uuid4()),
                'content': body.get('content', ''),
                'tags': body.get('tags', []),
                'timestamp': body.get('timestamp', datetime.now().isoformat()),
                'created': datetime.now().isoformat()
            }
            memories.append(memory)
            save_memories(memories)
            return self.send_json({'success': True, 'memory': memory})

        if parsed.path == '/api/memory/export':
            return self.send_json({'success': True, 'memories': load_memories()})

        if parsed.path == '/api/memory/import':
            body = self.read_body()
            existing = load_memories()
            imported = body.get('memories', [])
            merged = existing + imported
            save_memories(merged)
            return self.send_json({'success': True, 'imported': len(imported), 'total': len(merged)})

        self.send_json({'error': 'Not found'}, 404)

    def do_DELETE(self):
        parsed = urlparse(self.path)
        if parsed.path.startswith('/api/memory/'):
            mem_id = parsed.path.split('/api/memory/')[1]
            memories = load_memories()
            before = len(memories)
            memories = [m for m in memories if m['id'] != mem_id]
            save_memories(memories)
            return self.send_json({'success': True, 'deleted': before - len(memories)})
        self.send_json({'error': 'Not found'}, 404)

    def log_message(self, format, *args):
        pass

    def log_error(self, format, *args):
        print('ERROR:', format % args)

if __name__ == '__main__':
    server = http.server.HTTPServer(('0.0.0.0', PORT), MemoryHandler)
    print(f'Federation Memory Server running on http://localhost:{PORT}')
    print(f'Dashboard: http://localhost:{PORT}/')
    server.serve_forever()
